File: src/ObjectManager/Object.Tes/FilePacks/Records/345-DOOR.Door.cs
using OA.Core;
using System;

namespace OA.Tes.FilePacks.Records
{
    public class DOORRecord : Record, IHaveEDID, IHaveMODL
    {
        public override string ToString() => $"DOOR: {EDID.Value}";
        public STRVField EDID { get; set; } // door ID
        public STRVField FULL; // door name
        public MODLGroup MODL { get; set; } // NIF model filename
        public FMIDField<SCPTRecord>? SCRI; // Script (optional)
        public FMIDField<SOUNRecord> SNAM; // Open sound
        public FMIDField<SOUNRecord> ANAM; // Close sound
        // TES4
        public FMIDField<SOUNRecord> BNAM; // Loop sound
        public BYTEField FNAM; // Flags
        public FMIDField<Record> TNAM; // Random teleport destination

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, uint dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "FULL": FULL = new STRVField(r, dataSize); return true;
                case "FNAM": if (formatId != GameFormatId.Tes3) FNAM = new BYTEField(r, dataSize); else FULL = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                case "SNAM": SNAM = new FMIDField<SOUNRecord>(r, dataSize); return true;
                case "ANAM": ANAM = new FMIDField<SOUNRecord>(r, dataSize); return true;
                case "BNAM": ANAM = new FMIDField<SOUNRecord>(r, dataSize); return true;
                case "TNAM": TNAM = new FMIDField<Record>(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-DOOR.Door.swift
//
//  DOORRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class DOORRecord: Record, IHaveEDID, IHaveMODL {
    public override var description: String { return "DOOR: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var FULL: STRVField! // Door name
    public var MODL: MODLGroup? = nil // NIF model filename
    public var SCRI: FMIDField<SCPTRecord>? = nil // Script (optional)
    public var SNAM: FMIDField<SOUNRecord>!  // Open Sound
    public var ANAM: FMIDField<SOUNRecord>!  // Close Sound
    // TES4
    public var BNAM: FMIDField<SOUNRecord>! // Loop Sound
    public var FNAM: BYTEField! // Flags
    public var TNAM: FMIDField<Record>! // Random teleport destination

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "FULL": FULL = r.readSTRV(dataSize)
        case "FNAM": if format != .TES3 { FNAM = r.readT(dataSize) } else { FULL = r.readSTRV(dataSize) }
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
        case "SNAM": SNAM = FMIDField<SOUNRecord>(r, dataSize)
        case "ANAM": ANAM = FMIDField<SOUNRecord>(r, dataSize)
        case "BNAM": ANAM = FMIDField<SOUNRecord>(r, dataSize)
        case "TNAM": TNAM = FMIDField<Record>(r, dataSize)
        default: return false
        }
        return true
    }
}
